src/platform/i386/vga.c

```diff
@@ -0,0 +1,211 @@
+/* The number of lines. */
+#define LINES                  25
+/* The attribute of an character. */
+#define ATTRIBUTE              0x0F
+/* The video memory address. */
+#define VIDEO                  0xB8000
+
+#define SPACE                  0x20
+#define STRLEN_MAX             (COLUMNS * 2)
+
+#define VGA_CTL_REG            0x3D4
+#define VGA_DATA_REG           0x3D5
+
+#define KEY_DEVICE             0x60
+#define KEY_PENDING            0x64
+
+/* Variables. */
+/* Save the X position. */
+static int csr_x;
+/* Save the Y position. */
+static int csr_y;
+/* Point to the video memory. */
+static volatile unsigned char *video;
+/* Attribute of a character */
+static int attrib = ATTRIBUTE;
+
+/* Forward declarations. */
+static unsigned short *memsetw(unsigned short *dest, unsigned short val, size_t count);
+static void puts(unsigned char *str);
+static void move_csr(void);
+static void cls(void);
+static void cll(void);
+static void putchar (int c);
+
+/* Utility: set 16bit memory with 16bit value */
+static unsigned short *memsetw(unsigned short *dest, unsigned short val, size_t count)
+{
+	unsigned short *temp = NULL;
+	/* Same as memset, but this time, we're working with a 16-bit
+	 *  'val' and dest pointer. Your code can be an exact copy of
+	 *  the memset, provided that your local variables if any, are
+	 *  unsigned short */
+	temp = (unsigned short *)dest;
+	for( ; count != 0; count--) *temp++ = val;
+	return dest;
+}
+
+/* Updates the hardware cursor: the little blinking line
+ *  on the screen under the last character pressed! */
+static void move_csr(void)
+{
+	unsigned pos = 0;
+
+	/* The equation for finding the index in a linear
+	 *  chunk of memory can be represented by:
+	 *  Index = [(y * width) + x] */
+	pos = csr_y * COLUMNS + csr_x;
+
+	/* This sends a command to indicies 14 and 15 in the
+	 *  CRT Control Register of the VGA controller. These
+	 *  are the high and low bytes of the index that show
+	 *  where the hardware cursor is to be 'blinking'. To
+	 *  learn more, you should look up some VGA specific
+	 *  programming documents. A great start to graphics:
+	 *  http://www.brackeen.com/home/vga */
+	outb(VGA_CTL_REG, 14);
+	outb(VGA_DATA_REG, pos >> 8);
+	outb(VGA_CTL_REG, 15);
+	outb(VGA_DATA_REG, pos);
+}
+
+/* Clears the line from current y position */
+static void cll(void)
+{
+	unsigned blank = 0;
+	/* Again, we need the 'short' that will be used to
+	 *  represent a space with color */
+	blank = SPACE | (attrib << 8);
+	/* Sets the entire screen to spaces in our current
+	 *  color */
+	memsetw ((unsigned short *)video + csr_y * COLUMNS + csr_x, blank, COLUMNS - csr_x);
+}
+
+
+/* Clears the screen */
+static void cls(void)
+{
+	unsigned blank = 0;
+	int i = 0;
+
+	/* Again, we need the 'short' that will be used to
+	 *  represent a space with color */
+	blank = SPACE | (attrib << 8);
+
+	/* Sets the entire screen to spaces in our current
+	 *  color */
+	for(i = 0; i < LINES; i++)
```

Comment:
Always use `{}` unless the body is on the same line as the conditional.  Style guide.